api/v1alpha4/clusterclass_webhook.go

```diff
@@ -132,6 +161,107 @@ func (in *ClusterClass) validate(old *ClusterClass) error {
 			)
 		}
 	}
+	return allErrs
+}
+
+func (in ClusterClass) validateRefTemplates() field.ErrorList {
+	var allErrs field.ErrorList
+
+	if !strings.HasSuffix(strings.ToLower(in.Spec.Infrastructure.Ref.Kind), "template") {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec", "infrastructure", "ref", "kind"),
+				in.Spec.Infrastructure.Ref.Kind,
+				"must be a template",
+			),
+		)
+	}
+	if !strings.HasSuffix(strings.ToLower(in.Spec.ControlPlane.Ref.Kind), "template") {
```

Comment:
I think it would be nice to move this repeating validation into a func (and then call it for every ref). If I see correctly, you can hand over the ref and the path and everything else is the same. Same for apiVersion. I think this will make the implementation more concise.

Reply:
This will be taken care of by with https://github.com/kubernetes-sigs/cluster-api/pull/4992#discussion_r675096335